### PR TITLE
Supplementary queries: allow plugin decoupling by allowing providers to return a request instance

### PR DESCRIPTION
**Pull request by matyax**
Currently, plugins that want to support Logs Volume and/or Logs Sample need to import dependencies from `logsModel`:

imagen

### Proposed approach

1. Deprecate `getDataProvider(): Observable<DataQueryResponse>`.

By the way this function was designed, it created a dependency between data sources and core, where `queryLogsSample`, `queryLogsVolume` needed to be needlessly imported from core. 

2. Introduce `getSupplementaryRequest(): DataQueryRequest<TQuery>`. 

Observing on the three (Loki, Elasticsearch, ClickHouse) implementations of supplementary queries, it was evident that inversion of control was possible, and `queryLogsSample`, `queryLogsVolume` could be called from core Grafana by just passing a `DataQueryRequest` instance to the data sources and getting another `DataQueryRequest` for the supplementary queries.

Thus, `getSupplementaryRequest` has been introduced to receive and return a `DataQueryRequest` instance.

**Why do we need this feature?**

To allow externalization of Loki and Elasticsearch.

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/grafana/issues/72632 and https://github.com/grafana/grafana/issues/72631

# Deprecation notice

For data sources that implement `DataSourceWithSupplementaryQueriesSupport`, `getDataProvider()` has been deprecated in favor of getSupplementaryRequest()`.

### CODE REVIEW

**Review by ivanahuckova — COMMENTED** [review]
I will do a more proper review, but I think we should keep the API simple and replace `Observable<DataQueryResponse>` with  `DataQueryRequest<TQuery>` because `Observable<DataQueryResponse>` won't be used by any of the existing data sources and for plugin developers, having `DataQueryRequest<TQuery>` would be preferred.

**Comment by ivanahuckova** [comment]
@ifrost could you please have a look at this one and share your thoughts as you've created this API. 🙂

**Comment by matyax** [comment]
Sounds good to me @ivanahuckova !

**Review by mckn — COMMENTED** [review]
LGTM! But I think @ifrost is the best man to review this PR and give final approval 👍🏻

**Comment by matyax** [comment]
@sasklacz I recall you were trying to make logs volume work somewhere, would this change affect that?

**Review by ivanahuckova — COMMENTED** [review]
Overall, I like changes and direction. But I discovered some issue related to parsing level. The logs volume doesn't load then, due to issue when parsing level.

https://github.com/grafana/grafana/assets/30407135/d8c4ae7f-3dfd-4a46-97a8-d0198cfa0fec

File: packages/grafana-data/src/types/logs.ts [hunk 1]
```diff
@@ -229,7 +229,7 @@ export interface DataSourceWithSupplementaryQueriesSupport<TQuery extends DataQu
   getDataProvider(
     type: SupplementaryQueryType,
     request: DataQueryRequest<TQuery>
-  ): Observable<DataQueryResponse> | undefined;
+  ): Observable<DataQueryResponse> | DataQueryRequest<TQuery> | undefined;
```
[the remark this answers is not in the document]
Comment (matyax) [reply packages/grafana-data/src/types/logs.ts:232]: Makes sense! Done

File: public/app/features/logs/logsModel.ts [hunk 2]
```diff
@@ -611,15 +611,31 @@ type LogsVolumeQueryOptions<T extends DataQuery> = {
...
+}
+
 /**
  * Creates an observable, which makes requests to get logs volume and aggregates results.
  */
 export function queryLogsVolume<TQuery extends DataQuery, TOptions extends DataSourceJsonData>(
   datasource: DataSourceApi<TQuery, TOptions>,
   logsVolumeRequest: DataQueryRequest<TQuery>,
-  options: LogsVolumeQueryOptions<TQuery>
+  options?: LogsVolumeQueryOptions<TQuery>
```
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/logs/logsModel.ts:633]: Yes, range and targets are required for re-using results on zoom-in: https://github.com/grafana/grafana/blob/aa670280fc46b2d96e3096f1395e5145c715cb44/public/app/features/explore/state/query.ts#L878

File: public/app/features/logs/logsModel.ts [hunk 3]
```diff
@@ -611,15 +611,31 @@ type LogsVolumeQueryOptions<T extends DataQuery> = {
   range: TimeRange;
 };
 
+function defaultExtractLevel(dataFrame: DataFrame): LogLevel {
```
Comment (ifrost) [inline public/app/features/logs/logsModel.ts:614]: This logic may not be applicable to all data sources. I wonder if we may need to be more flexible to allow overriding it as well. At least we should support data plane spec (which I think should read level from `severity`), this way data sources could ensure the level is extracted correctly with proper data frame format.

Comment (ifrost) [reply public/app/features/logs/logsModel.ts:614]: Maybe log volume response should also be part of data plane? 🤔

Comment (matyax) [reply public/app/features/logs/logsModel.ts:614]: It's applicable to the 2 data sources that currently support logs volume (Loki and Elasticsearch), and can also be overriden by `options`: https://github.com/grafana/grafana/pull/80281/files#diff-cda90ed4a0399053b7808392e0fb6b8f03ed64289acecb0978dc0ba544dd40f8R637

File: public/app/features/logs/logsModel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/features/logs/logsModel.ts:633]: @ifrost I really don't see where `range` and `targets` are used for zoom-in, could you please point me to where are they used?

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 4]
```diff
@@ -553,10 +546,6 @@ export class ElasticDatasource
   }
 
   getSupplementaryQuery(options: SupplementaryQueryOptions, query: ElasticsearchQuery): ElasticsearchQuery | undefined {
-    if (!this.getSupportedSupplementaryQueryTypes().includes(options.type)) {
-      return undefined;
-    }
-
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:559]: Why is this removed?

Comment (matyax) [reply public/app/plugins/datasource/elasticsearch/datasource.ts:559]: There's no need to add that, when in both cases the switch is already doing that work:

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/elasticsearch/datasource.ts#L626

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L249

File: public/app/features/logs/logsModel.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/logs/logsModel.ts:619]: Hm... but to override `options` the plugin would need to call `queryLogsVolume` from `logsModel` which wouldn't have access to, right? So there won't be an easy way to override options in plugins code 🤔

Comment (matyax) [reply public/app/features/logs/logsModel.ts:617]: I think the key here is _may not_. So far we have 2 implementations using logs volume, and both share the same level extraction logic. For now I would say YAGNI. What do you think?

File: public/app/features/logs/logsModel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/features/logs/logsModel.ts]: It took me so long to figure out why it was not working, because this not null assertion operator (!) was hiding a potential bug: https://github.com/grafana/grafana/blob/main/public/app/features/explore/utils/supplementaryQueries.ts#L82

**Comment by matyax** [comment]
The PR as it is currently, allows to decouple logs model from Elastic and Loki datasources. The main item pending discussion is what to do with `extractLevel`. Like I [mentioned in a comment](https://github.com/grafana/grafana/pull/80281#discussion_r1458834646), both data sources using these APIs don't require a special flavor of `extractLevel`. I would personally not make it customizable for now, until we have a concrete case that we need to support. What do you think @ifrost @ivanahuckova ?

**Comment by ivanahuckova** [comment]
> I would personally not make it customizable for now, until we have a concrete case that we need to support. What do you think @ifrost @ivanahuckova ?

sounds good to me! Let's keep it DRY and introduce it when needed. 🙂

File: public/app/features/logs/logsModel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/logs/logsModel.ts]: >  I don't see the point of using options.range. Same for options.targets.

`options.targets` was needed to pass the targets from the original requests not targets of the log volume requests. 

`options.range` was needed in the past when a partial log volume was rendered and the returned range didn't match the request range. But we moved the implementation of partial log volume out of data sources so maybe we don't need it anymore 🤔

File: public/app/features/logs/logsModel.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/logs/logsModel.ts:646]: I'm just concerned about backwards compatibility, and a dreadful maintenance of multiple returned results if we add one more on top of it when we need it... Do you need getProvider() API to be public or we will keep it internal? I'm also now wondering if we need to support handling Observable returned from `getProvider` as it seems Elastic and Loki will always return a request now 🤔  In theory we could rename `getProvider` to `mapToSupplementaryRequest` 🤔

Comment (matyax) [reply public/app/features/logs/logsModel.ts:646]: > In theory we could rename getProvider to mapToSupplementaryRequest 🤔

+1 absolutely.

File: public/app/features/logs/logsModel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/features/logs/logsModel.ts]: > options.targets was needed to pass the targets from the original requests not targets of the log volume requests.

Yeah! I eventually saw this, and updated the implementation to allow reuse of supplementary queries.

> But we moved the implementation of partial log volume out of data sources so maybe we don't need it anymore 🤔

It seems that we don't. `targets` was being modified inside `getDataProvider`, but range was always === to the original one.

File: public/app/features/logs/logsModel.ts (hunk 3 above)
Comment (ivanahuckova) [reply public/app/features/logs/logsModel.ts:646]: I think this logic will actually be applicable for all data sources, that return correct data frames in  `Time Series Wide Format` or `Time Series Multi Format` with `level` label. Maybe instead of `extractLevel` function we just need the way to override `level` label, if it does not have name `level`. That way data sources can override getting of level, and we support it correctly for every data source that produces valid data plane compatible frames. 

And considering log volume hasn't worked for any data source besides Loki and Elasticsearch, we don't need to worry about backward compatibility.

Comment (ivanahuckova) [reply public/app/features/logs/logsModel.ts:646]: > I'm also now wondering if we need to support handling Observable returned from getProvider as it seems Elastic and Loki will always return a request now 🤔

I think this was actually removed and we only expect `DataQueryRequest<TQuery>` and not observable.

```
getDataProvider(
    type: SupplementaryQueryType,
    request: DataQueryRequest<TQuery>
  ): DataQueryRequest<TQuery> | undefined;
  ```

Comment (matyax) [reply public/app/features/logs/logsModel.ts:646]: `getSupplementaryQueryRequest` sounds great.

File: public/app/features/logs/logsModel.ts [hunk 5]
```diff
@@ -640,11 +639,23 @@ const updateLogsVolumeConfig = (
 };
 
 type LogsVolumeQueryOptions<T extends DataQuery> = {
-  extractLevel: (dataFrame: DataFrame) => LogLevel;
+  extractLevel?: (dataFrame: DataFrame) => LogLevel;
```
Comment (ivanahuckova) [inline public/app/features/logs/logsModel.ts:642]: Just to confirm, what is the decision related to logs volume query options and level overrided:
- we don't need it for now
- we want to have it as function
- instead of function, we want to just have the option to specify the name of `level` field (e.g. somewhere it can be `error_level`, or `LVL`, or `LEVEL`  or anything else?

Comment (matyax) [reply public/app/features/logs/logsModel.ts:642]: Both `we don't need it for now` and `option to specify the level field` works for me. Now that we're at it, we can easily add support to specify the `level` field and win-win.

File: public/app/features/logs/logsModel.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/logs/logsModel.ts:646]: Apart from Loki and Elasticsearch there's one more internal implementation in [ClickHouse Datasource](https://github.com/grafana/clickhouse-datasource/blob/main/src/data/CHDatasource.ts#L63). 

We need to keep the old method, though it can be marked as deprecated. We ccan introduce a new called `getSupplementaryQueryRequest`. ClickHouse should be updated and only then the old method can be removed, but a big question is how many versions back we need to support 🤔  What are the latest guidelines for internal plugins? Do you own the ClickHouse plugin as well? Would it be possible to use the new `getSupplementaryQueryRequest` and the current level extraction logic in that plugin? 🤔

Comment (matyax) [reply public/app/features/logs/logsModel.ts:646]: Maybe we could send a PR to update ClickHouse with these changes? I think doing in two steps would be the longer path just for one data source. I wouldn't mind at all doing that as a follow up to the API changes here. What do you all think? Nice catch btw, we almost miss it!

Comment (ifrost) [reply public/app/features/logs/logsModel.ts:646]: It depends on when new version of a plugin is released and how many previous versions of Grafana we need to support. If we release the change in the API before plugin is updated it will stop working. Also, I think plugins are not updated automatically, so I guess there's need to be some grace period with backwards compatibility anyway. What guidelines do you have for newly externalized plugins?

**Comment by matyax** [comment]
@ivanahuckova @ifrost Updated according to our recent conversation.

File: packages/grafana-data/src/types/logs.ts [hunk 6]
```diff
@@ -225,36 +225,44 @@ export interface DataSourceWithSupplementaryQueriesSupport<TQuery extends DataQu
...
    * supplementary query type and original request.
+   * @deprecated Use getSupplementaryQueryRequest() instead
    */
-  getDataProvider(
+  getDataProvider?(
     type: SupplementaryQueryType,
     request: DataQueryRequest<TQuery>
   ): Observable<DataQueryResponse> | undefined;
+  /**
+   * Given a DataQueryRequest returns a new DataQueryRequest to be used to fetch supplementary data.
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/logs.ts:235]: I don't understand this first part of the comment.

File: public/app/features/explore/utils/supplementaryQueries.ts [hunk 7]
```diff
@@ -130,7 +130,19 @@ export const getSupplementaryQueryProvider = (
...
+      if (datasource.getDataProvider) {
+        return datasource.getDataProvider(type, dsRequest);
+      } else if (datasource.getSupplementaryRequest) {
+        const provider = datasource.getSupplementaryRequest(type, dsRequest);
+        if (!provider) {
+          return provider;
+        }
+        return type === SupplementaryQueryType.LogsVolume
+          ? queryLogsVolume(datasource, provider, { targets: dsRequest.targets })
+          : queryLogsSample(datasource, provider);
```
Comment (ivanahuckova) [inline public/app/features/explore/utils/supplementaryQueries.ts:142]:
```suggestion
        const request = datasource.getSupplementaryRequest(type, dsRequest);
        if (!provider) {
          return undefined;
        }
        return type === SupplementaryQueryType.LogsVolume
          ? queryLogsVolume(datasource, request, { targets: dsRequest.targets })
          : queryLogsSample(datasource, request);
```

Comment (ivanahuckova) [reply public/app/features/explore/utils/supplementaryQueries.ts:142]: I think distinguishing between provider and request makes this code more readable.

File: public/app/features/explore/utils/supplementaryQueries_legacy.test.ts [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+import { DataQuery } from '@grafana/schema';
+
+import { MockDataSourceApi } from '../../../../test/mocks/datasource_srv';
+import { MockDataQueryRequest, MockQuery } from '../../../../test/mocks/query';
+import { ExplorePanelData } from '../../../types';
+import { mockExplorePanelData } from '../__mocks__/data';
+
+import { getSupplementaryQueryProvider } from './supplementaryQueries';
+
+class MockDataSourceWithSupplementaryQuerySupport
```
Comment (ivanahuckova) [inline public/app/features/explore/utils/supplementaryQueries_legacy.test.ts:29]: I would add a comment that this whole file should be removed when we remove support for `getDataProvider` in `DataSourceWithSupplementaryQueriesSupport`

File: public/app/plugins/datasource/elasticsearch/datasource.ts [hunk 9]
```diff
@@ -538,13 +534,15 @@ export class ElasticDatasource
     }
   };
 
-  getDataProvider(
+  /**
+   * Implemented for DataSourceWithSupplementaryQueriesSupport.
+   * It generates a DataQueryRequest for a specific supplementary query type.
+   * @returns A DataQueryRequest for the supplementary queries or undefined if not supported.
+   */
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/elasticsearch/datasource.ts:541]: 🤩

**Review by ivanahuckova — COMMENTED** [review]
Left couple of comments, nothing blocking. Really nice job with this one!

**Review by ivanahuckova — APPROVED** [review]
I wanted to approve, but accidentally selected "Comment" 🙈

File: packages/grafana-data/src/types/logs.ts (hunk 6 above)
Comment (matyax) [reply packages/grafana-data/src/types/logs.ts:235]: I mean with this that you receive a `DataQueryRequest` and return another `DataQueryRequest` to fetch the supplementary data. Is there a specific part of the comment that you find particularly unclear?

If it's the "given" part, it's like saying "given an array of numbers, returns odd numbers"

File: public/app/features/explore/utils/supplementaryQueries_legacy.test.ts (hunk 8 above)
Comment (matyax) [reply public/app/features/explore/utils/supplementaryQueries_legacy.test.ts:29]: Done

**Comment by ifrost** [comment]
> Observing on the three (Loki, Elasticsearch, ClickHouse) implementations of supplementary queries, it was evident that inversion of control was possible, and queryLogsSample, queryLogsVolume could be called from core Grafana by just passing a DataQueryRequest instance to the data sources and getting another DataQueryRequest for the supplementary queries.

@grafana/observability-logs - have you checked with ClickHouse developers if they could have a look into implementing new API?

**Comment by matyax** [comment]
No, thanks for the reminder @ifrost !

**Comment by matyax** [comment]
Done now.
